### PR TITLE
Type consistency enforcement Qos

### PR DESCRIPTION
**Pull request by sonndinh**
Updated:
- Type assignability to support `ignore_member_names`
- Type assignability unit tests
- Parameter list converter, discovery base, sedp and related parts to support TypeConsistencyEnforcementQosPolicy
- XTypes end-to-end tests to test the added Qos policy

Out-of-scope (future work):
- The 3 remaining booleans of TypeConsistencyEnforcementQosPolicy (`prevent_type_widening`, `ignore_sequence_bounds`, `ignore_string_bounds`) are not yet implemented and currently taking default values. Setting new values to them doesn't change the behavior of type assignability and matching.

### CODE REVIEW

**Review by kuznetsovmoci — CHANGES_REQUESTED** [review]
Do we need code in Sedp.cpp, ParameterListConverter.cpp, etc. to support new QoS type_consistency parameter?

**Comment by sonndinh** [comment]
> Do we need code in Sedp.cpp, ParameterListConverter.cpp, etc. to support new QoS type_consistency parameter?

I'm not sure about `Sedp`, but probably `ParameterListConverter` will need to be updated.

File: bin/dcps_tests.lst [hunk 1]
```diff
@@ -426,7 +426,7 @@ tests/FACE/Compiler/idl_test1_main/run_test.pl: !DCPS_MIN !WCHAR !NO_BUILT_IN_TO
 tests/FACE/Compiler/idl_test3_main/run_test.pl: !DCPS_MIN !WCHAR !NO_BUILT_IN_TOPICS
 tests/FACE/Compiler/idl_test_fixed/run_test.pl: !DCPS_MIN !WCHAR !NO_BUILT_IN_TOPICS
 
-tests/DCPS/XTypes/run_test.pl: !DCPS_MIN
+tests/DCPS/XTypes/run_test.pl --type-consistency: !DCPS_MIN
```
Comment (mitza-oci) [inline bin/dcps_tests.lst:429]: Does it need to have the `--type-consistency` command line argument if that's the only way it's ever invoked?

File: dds/DCPS/RTPS/Sedp.cpp [hunk 2]
```diff
@@ -4798,24 +4799,32 @@ Sedp::populate_discovered_reader_msg(
...
+
   drd.contentFilterProperty.contentFilteredTopicName =
     OPENDDS_STRING(DCPS::GuidConverter(subscription_id)).c_str();
   drd.contentFilterProperty.relatedTopicName = topic_name.c_str();
   drd.contentFilterProperty.filterClassName = ""; // PLConverter adds default
   drd.contentFilterProperty.filterExpression = sub.filterProperties.filterExpression;
   drd.contentFilterProperty.expressionParameters = sub.filterProperties.expressionParameters;
+
+  const CORBA::ULong len = drd.readerProxy.associatedWriters.length();
+  const CORBA::ULong added_len = sub.remote_expectant_opendds_associations_.size();
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.cpp:4818]: See warnings reported by GitHub Actions on this line

File: dds/DCPS/XTypes/TypeAssignability.h [hunk 3]
```diff
@@ -20,18 +20,68 @@ namespace XTypes {
...
+    : tl_service_(tls)
+  {
+    // Use default type consistency values
+    type_consistency_.prevent_type_widening = false;
+    type_consistency_.ignore_sequence_bounds = true;
+    type_consistency_.ignore_string_bounds = true;
+    type_consistency_.ignore_member_names = false;
+  }
+
+  explicit TypeAssignability(TypeLookupService_rch tls,
```
Comment (mitza-oci) [inline dds/DCPS/XTypes/TypeAssignability.h:47]:
```suggestion
  TypeAssignability(TypeLookupService_rch tls,
```

File: tests/DCPS/XTypes/Pub/XTypesPublisher.cpp [hunk 4]
```diff
@@ -1,163 +1,245 @@
...
 void write_plain_cdr_struct(const DataWriter_var& dw)
 {
   PlainCdrStructDataWriter_var pdw = PlainCdrStructDataWriter::_narrow(dw);
 
   PlainCdrStruct pcs;
   pcs.key = key_value;
   pcs.value = 1;
-  pdw->write(pcs, HANDLE_NIL);
+
+  ReturnCode_t ret = pdw->write(pcs, HANDLE_NIL);
```
Comment (mitza-oci) [inline tests/DCPS/XTypes/Pub/XTypesPublisher.cpp:14]:
```suggestion
  const ReturnCode_t ret = pdw->write(pcs, HANDLE_NIL);
```

File: tests/DCPS/XTypes/Pub/XTypesPublisher.cpp [hunk 5]
```diff
@@ -182,16 +264,19 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
   DomainParticipant_var dp;
   create_participant(dpf, dp);
   if (!dp) {
-    ACE_ERROR((LM_ERROR, "ERROR: create_participant() failed"));
+    ACE_ERROR((LM_ERROR, "ERROR: create_participant failed"));
```
Comment (mitza-oci) [inline tests/DCPS/XTypes/Pub/XTypesPublisher.cpp:267]:
```suggestion
    ACE_ERROR((LM_ERROR, "ERROR: create_participant failed\n"));
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**